File: ch05/38/GetNumberOfK.py
# -*- coding:utf-8 -*-
class Solution:
    def GetNumberOfK(self, data, k):
        # write code here
        if len(data) == 0:
            return 0;

        startK = self.GetFirstK(data, k) # 返回重复数字出现的第一个位置
        endK = self.GetLastK(data, k)
        
        print "start:", startK
        print "end:",endK
        if startK == -1 or endK == -1:
            return 0
        
        num = endK - startK + 1        
        return num
    
    def GetFirstK(self, data, k):
        if len(data) == 0:
            return -1
        
        start = 0
        end = len(data) - 1
        while start <= end:
            mid = (start + end) >> 1
            if data[mid] == k:
                if mid == 0 or (mid > 0 and data[mid - 1] != k):
                    return mid
                else:
                    end = mid - 1
            elif data[mid] < k:
                start = mid + 1
            else:
                end = mid - 1
        return -1
    
    def GetLastK(self, data, k):
        if len(data) == 0:
            return -1

        start = 0
        end = len(data) - 1
        while start <= end:
            mid = (start + end)>>1
            if data[mid] == k:
                if mid == len(data) - 1 or (mid < len(data) and data[mid + 1] != k):
                    return mid
                else:
                    start = mid + 1
            elif data[mid] < k:
                start = mid + 1
            else:
                end = mid - 1
        return -1

if __name__ == "__main__":
    # data = [1,2,4,4,4,5,6,7]
    # data = [1,2,3,3,3,3]
    data = [3,3,3,3,4,5]
    solu = Solution()
    print "k出现的次数为：", solu.GetNumberOfK(data, 3)
                
                

File: ch05/38/GetNumberOfK.cpp
// 统计一个数字在排序数组中出现的次数
//
class Solution {
public:
    int getFirstNumber(vector<int> data, int start, int end, int k){//返回出现第一次的下标
    	if(start > end)
            return -1;
        
        int middle = (end + start)/2;//注意求中间位置的公式
        int middleData = data[middle];
        
        if( middleData == k){
            if(middle == 0 || (middle > 0 && data[middle-1] != k))
            	return middle; 
            else 
                end = middle -1;
        }
        else if(middleData > k)
            end = middle -1;
        else 
            start = middle + 1; 
        return getFirstNumber(data, start, end, k);
    }
    
    int getLastNumber(vector<int> data, int start, int end, int k){
        if(start > end)
            return -1;
        int len = data.size();
        //int result = 0;
        int middle = (end + start)/2;
        int middleData = data[middle];
        
        if(middleData == k){
            if(middle == len - 1 || (data[middle+1] != k && middle < len - 1)) 
                return middle;
            else 
                start = middle + 1;
        }
        else if(middleData < k)
            start = middle + 1;
        else 
            end = middle -1;
    	
        return getLastNumber(data, start, end, k);;
    }
    
    int GetNumberOfK(vector<int> data ,int k) {
        int number = 0;
        int len = data.size();
        if(len == 0)
            return 0;
        
        int first = getFirstNumber(data, 0, len - 1, k);
        int last = getLastNumber(data, 0, len-1, k);
        
        if(first > -1 && last > -1)
            number = last - first + 1;
        
        return number;
    }
};